File: 240426/가장 작은 x 찾기/find-smallest-x.py
import math

N = int(input())
ab_lst = [tuple(map(int, input().split())) for _ in range(N)]

x_lst = []
for pair_idx, (a, b) in enumerate(ab_lst, start = 1):
    x_min, x_max = math.ceil(a / 2 ** pair_idx), math.ceil(b / 2 ** pair_idx)
    x_lst.append((x_min, x_max))

x_min_max, x_max_min = max(list(zip(*x_lst))[0]), min(list(zip(*x_lst))[1])
answers = list(range(x_min_max, x_max_min + 1))
print(answers[0])

File: 260323/가장 작은 x 찾기/find-smallest-x.cpp
#include <bits/stdc++.h>
using namespace std;

long long ceil_div(long long a, long long b) 
{
    return (a + b - 1) / b;
}

int main() 
{
    int N;
    cin >> N;

    vector<pair<long long, long long>> ab_lst(N);
    for (int i = 0; i < N; i++) 
    {
        cin >> ab_lst[i].first >> ab_lst[i].second;
    }

    vector<pair<long long, long long>> x_lst;

    for (int pair_idx = 1; pair_idx <= N; pair_idx++) 
    {
        long long a = ab_lst[pair_idx - 1].first;
        long long b = ab_lst[pair_idx - 1].second;
        long long div = 1LL << pair_idx;

        long long x_min = ceil_div(a, div);
        long long x_max = ceil_div(b, div);

        x_lst.push_back({x_min, x_max});
    }

    long long x_min_max = x_lst[0].first;
    long long x_max_min = x_lst[0].second;

    for (int i = 1; i < N; i++) 
    {
        x_min_max = max(x_min_max, x_lst[i].first);
        x_max_min = min(x_max_min, x_lst[i].second);
    }

    cout << x_min_max << '\n';
    return 0;
}